### PR TITLE
feat(list)!: add list navigation methods (first, last, previous, next)

### PR DESCRIPTION
**Pull request by joshka**
Also cleans up the list example significantly (see also
<https://github.com/ratatui-org/ratatui/issues/1157>)
    
Fixes: <https://github.com/ratatui-org/ratatui/pull/1159>
    
BREAKING CHANGE: The `List` widget now clamps the selected index to the
bounds of the list when navigating with `first`, `last`, `previous`, and
`next`, as well as when setting the index directly with `select`.

### CODE REVIEW

File: src/symbols.rs [hunk 1]
```diff
@@ -470,6 +470,28 @@ pub mod border {
...
+    pub const EMPTY: Set = Set {
+        top_left: " ",
+        top_right: " ",
+        bottom_left: " ",
+        bottom_right: " ",
+        vertical_left: " ",
+        vertical_right: " ",
+        horizontal_top: " ",
+        horizontal_bottom: " ",
+    };
```
Comment (joshka) [inline src/symbols.rs:494]: This likely needs to go as a different PR.

I needed it because a title on a block without borders set doesn't apply the border style to the line that the title is on. This seems odd, but adding an empty / solid border set makes simple work of that. (I started with solid and then flipped to empty and didn't delete the code.)

File: src/widgets/list.rs [hunk 2]
```diff
@@ -1301,7 +1472,7 @@ mod tests {
             &single_item,
             Some(1),
             [
-                "  Item 0  ",
+                ">>Item 0  ",
```
Comment (joshka) [inline src/widgets/list.rs:1475]: This is a change that's likely more correct than the existing approach, but is probably a breaking change and should be noted as such.

File: src/widgets/list.rs [hunk 3]
```diff
@@ -156,6 +156,72 @@ impl ListState {
...
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use ratatui::{prelude::*, widgets::*};
+    /// let mut state = ListState::default();
+    /// state.select_next();
+    /// ```
+    pub fn select_next(&mut self) {
+        let next = self.selected.map_or(0, |i| i.saturating_add(1));
```
Comment (joshka) [inline src/widgets/list.rs:173]: It might be a good idea to add support for wrap around here and on prev, but for now I left that out of the PR.

File: examples/list.rs
Comment (joshka) [inline examples/list.rs:1]: This is mostly a rewrite

- Terminal related methods moved to the bottom
- pretty much everything rearranged in a more natural reading order (main, structs, App::default, List initialization, App logic, rendering logic
- Simplified the rendering a bunch
- code reduction of ~35 lines (64 if you don't count the small methods to navigate the todo list - these could be inline, but I think it generally makes sense to extract them)

File: src/widgets/list.rs (hunk 2 above)
Comment (a-kenji) [reply src/widgets/list.rs:1475]: Agreed, in my lists I do the same by default. This feels much more natural. If consumers don't implement this manually - for me personally this always leads to confusing interactions.

File: src/symbols.rs (hunk 1 above)
Comment (joshka) [reply src/symbols.rs:494]: https://github.com/ratatui-org/ratatui/pull/1182

File: src/widgets/list.rs (hunk 3 above)
Comment (EdJoPaTo) [reply src/widgets/list.rs:173]: On my tree widgets I found wrap around to be a bit more confusing. Hitting keys with nothing selected selects something here. Up ends at the top. For someone to move to the extremes, Home and End seem way better.

Comment (joshka) [reply src/widgets/list.rs:173]: There are a variety of use cases where this is reasonable. What I meant was that this should be something that can be enabled by an app rather than something that needs to be part of this PR.

**Comment by joshka** [comment]
Rebased on main, added breaking changes docs, updated commit message and PR comment to note breaking change.

**Review by orhun — APPROVED** [review]

**Comment by errwip** [comment]
I don't know if this is the right place to comment, but the clamping is incorect. It clamps to `items.len()` instead of `items.len() - 1`
It doesn't cause any problems for just displaying and navigating the list, BUT if you couple some other part of your code with `list_state.selected()`, you will get index out of bound errors.

My current workaround:
```
let next = self.list_state.selected().map_or(0, |i| i+1);
self.list_state.select(Some(next.min(self.items.len()-1)));
```

**Comment by joshka** [comment]
Great find - mind submitting a PR?
